group_actions.py

```diff
@@ -101,6 +104,82 @@ def whitelist_package_in_group_devices(package_name):
+in environemtn. Return None if not present.
+"""
+def get_app_id(package_name, version_code):
+    app_api_instance = esperclient.ApplicationApi(esperclient.ApiClient(CONFIGURATION))
+    response = app_api_instance.get_all_applications(ENTERPRISE_ID)
+    #print(response)
+    i = 0
+
+    while (i < response.count):
+        # found the package already present in the environment
+        if response.results[i].package_name == package_name:
+            version_len = (len(response.results[i].versions))
+
+            # if version_code is none, then use the first version provided to install
+            j = 0
+            while (j < version_len and version_code is not None):
+                if response.results[i].versions[j].version_code == version_code:
+                    # found the same version code to install, store the j value to use
+                    # else use the first instance of App version given
+                    break
+                else:
+                    j = j+1
+
+            if j == version_len:
+                # this means that specific version code is not found in the list, do nothing
+                # bail out from here
+                print("Version not found")
+                return None
+
+            # Here will come if version found or no version is provided to install
+            app_version_id = response.results[i].versions[j].id
+            return app_version_id
+        else:
+            # check for next package in list
+            i = i+1
+    return None
+
+
+"""
+Run installation of app package name and specific app_version to device
+"""
+def run_install_command(device, package_name, app_version_id):
+
+    api_instance = esperclient.CommandsApi(esperclient.ApiClient(CONFIGURATION))
+    command_args = esperclient.CommandArgs(package_name=package_name, app_version=app_version_id)
+    command = esperclient.CommandRequest(command='INSTALL', command_args=command_args)
+    try:
+        api_response = api_instance.run_command(ENTERPRISE_ID, device.id, command)
+        if api_response.id is not None:
+            print('Install command fired successfully on ' + device.device_name + \
+                  ' for ' + package_name)
+    except ApiException as api_exception:
+        print("Exception when calling CommandsApi->run_command: %s\n" % api_exception)
+
+"""
+install package to entire group active devices
+"""
+def install_package_to_group_devices(package_name, version_code):
+
+    # get app_id from version code
+    app_id = get_app_id(package_name, version_code)
+    if app_id is None:
+        print("No Such App or App Version exist in Environment, Please upload it first and then install")
+        return
+
+    print("Install Currently active devices in group: ")
+    print([device.device_name for device in ACTIVE_DEVICE_LIST])
+
+    # install the app_id on all active devices
+    for device in ACTIVE_DEVICE_LIST:
+        run_install_command(device, package_name, app_id)
```

Comment:
This is fire and forget. Will it be a good idea to maintain result set of success and failure on device. How will one know on which all devices command succeeded ?

Reply:
Yes you are right . This is fire and forget. We can't wait for device to return back on command line. We can just say fire a command. May be we can add as an enchancement later